src/vcpu.rs

```diff
@@ -0,0 +1,222 @@
+    /// - Control Registers: CR0, CR2, CR3, CR4, CR8
+    /// - Additional: EFER, APIC base
+    /// - Interrupt_bitmap: Bitmap of pending external interrupts (KVM only, may
+    ///   be unused by vCPU implementations of other VMMs)
+    #[cfg(any(target_arch = "x86", target_arch = "x86_64"))]
+    fn set_sregs(&self, sregs: &SpecialRegisters) -> Result<()>;
+
+    /// Reads the floating point state from the vCPU
+    ///
+    /// - Floating Point MMX Registers 0-7
+    /// - Floating Pointer Control, Status, Tag Registers (FCW, FSW, FTW)
+    /// - Floating point exception state (Last FIP, FOP, FCS, FDS, FDP)
+    /// - XMM Registers 0-15
+    /// - MXCSR Control and Status Register
+    #[cfg(any(target_arch = "x86", target_arch = "x86_64"))]
+    fn get_fpu(&self) -> Result<FpuState>;
+
+    /// Writes the floating point state to the vCPU
+    ///
+    /// - Floating Point MMX Registers 0-7
+    /// - Floating Pointer Control, Status, Tag Registers (FCW, FSW, FTW)
+    /// - Floating point exception state (Last FIP, FOP, FCS, FDS, FDP)
+    /// - XMM Registers 0-15
+    /// - MXCSR Control and Status Register
+    #[cfg(any(target_arch = "x86", target_arch = "x86_64"))]
+    fn set_fpu(&self, fpu: &FpuState) -> Result<()>;
+
+    /// Defines the vCPU responses to the CPUID instruction
+    ///
+    /// The information contained in these responses can be set to be consistent
+    /// with hardware, kernel, userspace capabilities, and user requirements.
+    #[cfg(any(target_arch = "x86", target_arch = "x86_64"))]
+    fn set_cpuid2(&self, cpuid: &CpuId) -> Result<()>;
+
+    /// Read the Local APIC state from the vCPU
+    ///
+    /// LAPIC state is comprised of the memory-mapped Local APIC registers.
+    #[cfg(any(target_arch = "x86", target_arch = "x86_64"))]
+    fn get_lapic(&self) -> Result<LapicState>;
+
+    /// Write the Local APIC state to the vCPU
+    ///
+    /// LAPIC state is comprised of the memory-mapped Local APIC registers
+    #[cfg(any(target_arch = "x86", target_arch = "x86_64"))]
+    fn set_lapic(&self, klapic: &LapicState) -> Result<()>;
+
+    /// Read the Model-Specific Registers from the vCPU based on MSR indices
+    #[cfg(any(target_arch = "x86", target_arch = "x86_64"))]
+    fn get_msrs(&self, msrs: &mut MsrEntries) -> Result<(i32)>;
+
+    /// Write the Model-Specific Registers to the vCPU based on MSR indices
+    #[cfg(any(target_arch = "x86", target_arch = "x86_64"))]
+    fn set_msrs(&self, msrs: &MsrEntries) -> Result<()>;
+
+    /// Initialize a vCPU
+    ///
+    /// Specifies the CPU to present to the guest, and the optional features
+    /// it should have
+    #[cfg(any(target_arch = "arm", target_arch = "aarch64"))]
+    fn vcpu_init(&self, kvi: &VcpuInit) -> Result<()>;
+
+    /// Set a single register in the virtual CPU
+    #[cfg(any(target_arch = "arm", target_arch = "aarch64"))]
+    fn set_one_reg(&self, reg_id: u64, data: u64) -> Result<()>;
```

Comment:
Any way to read current value of a register on ARM/ARM64? Seems there's no equivalent get_one_reg() on ARM.
Could we combine set_one_reg() with set_regs() by defining StandardRegisters for ARM as :
struct StandardRegisters {
     reg_id: u64,
     data: u64,
}

Reply:
- What's the issue with implementing set_regs()/get_regs() on arm? Performance? Too many registers?
- Why not implement get_one_reg() / set_one_reg() on x86_64?

Reply:
I chose these to reach parity with what was provided by the kvm-ioctls crate, but it might have been overreaching for this PR since it's not something I'm familiar with. I saw an email fly by that someone is willing/able to do the ARM work for the kvm-ioctls crate, so maybe it makes sense to remove these until that work is done and then adapt when those are fleshed out.

While I can't speak for KVM, the get_one_reg()/set_one_reg() functionality is something that would be useful for Hyper-V. So maybe it does make sense that, if we keep it here, to have at least a conditional compilation for ARM or Windows, and let the WHP side implement it (with, as you suggest, a hypervisor-specific enum as an associated type). Or have it for all, and let KVM implement it with the `unimplemented!()` macro if there's no appropriate use for it.

Reply:
Just in case it's relevant, please consider using the Arm AArch64 Cortex-A bare-metal register access crate [here](https://github.com/rust-embedded/cortex-a).